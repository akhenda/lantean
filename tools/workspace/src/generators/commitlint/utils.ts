import { Tree } from '@nx/devkit';

import { NormalizedSchema } from './types';
import { CommitlintGeneratorSchema } from './schema';
import { defaultCommitlintConfigFile } from './constants';

import { getGitRepo } from '../../devkit';

/**
 * Normalize options for the commitlint generator.
 *
 * @param tree The virtual file system tree.
 * @param options The options passed to the generator.
 * @returns The normalized options.
 *
 * The normalized options include the following:
 * - `fileName`: The path to the commitlint configuration file. If not provided, the default value is {@link defaultCommitlintConfigFile}.
 * - `gitRepo`: The URL of the git repository.
 * - `appsDir`: The path to the apps directory.
 * - `libsDir`: The path to the libs directory.
 */
export function normalizeOptions(
  tree: Tree,
  options: CommitlintGeneratorSchema,
): NormalizedSchema {
  const fileName = options ? options.configFileName : defaultCommitlintConfigFile;

  return {
    ...options,
    fileName,
    gitRepo: getGitRepo(tree)?.replace(/^git@github\.com:/, 'https://github.com/'),
  };
}

/**
 * The content of the workspace README.md file.
 */
export const readmeContent = `
## Docs
[Commitlint & Semantic Release](./docs/commitlint.md)

## Run tasks
`;
