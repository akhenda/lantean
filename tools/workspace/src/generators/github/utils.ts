import { getWorkspaceLayout, Tree } from '@nx/devkit';

import { getGitRepo } from '../../devkit';

import { GithubGeneratorSchema, NormalizedSchema } from './schema';

export function normalizeOptions(
  tree: Tree,
  options: GithubGeneratorSchema
): NormalizedSchema {
  const layout = getWorkspaceLayout(tree);
  const appsDir = layout.appsDir === '.' ? 'apps' : layout.appsDir;
  const libsDir = layout.libsDir === '.' ? 'libs' : layout.libsDir;

  return {
    ...options,
    appsDir,
    libsDir,
    gitRepo: getGitRepo(tree)?.replace(
      /^git@github\.com:/,
      'https://github.com/'
    ),
  };
}
