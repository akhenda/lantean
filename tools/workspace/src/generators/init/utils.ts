import { InitGeneratorSchema, NormalizedSchema } from './schema';
import { defaultLintStagedConfigFile } from './constants';

/**
 * Normalize options for the init generator.
 *
 * @param options The options passed to the generator.
 * @returns The normalized options.
 *
 * The normalized options include the following:
 * - `fileName`: The path to the lint-staged configuration file. If not provided, the default value is {@link defaultLintStagedConfigFile}.
 */
export function normalizeOptions(
  options: InitGeneratorSchema
): NormalizedSchema {
  const fileName = options ? options.configFileName : defaultLintStagedConfigFile;

  return { ...options, fileName };
}
