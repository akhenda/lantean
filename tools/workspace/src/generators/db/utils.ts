import { getWorkspaceLayout, names, offsetFromRoot, Tree } from '@nx/devkit';

import { defaultLibDirectory, defaultLibName, defaultLibTags } from './constants';
import { NormalizedSchema, DbGeneratorSchema } from './schema';

import { NormalizedSchema as LoggingLibNormalizedSchema } from '../logging/schema';
import { getImportPath, getNpmScope } from '../../utils';

/**
 * Normalize options for the DB generator.
 *
 * @param tree The virtual file system tree.
 * @param options The options passed to the generator.
 * @param extra The extra options passed to the generator.
 * @returns The normalized options.
 *
 * The normalized options include the following:
 * - `name`: The name of the library.
 * - `projectName`: The name of the project.
 * - `projectRoot`: The path to the project root.
 * - `projectDirectory`: The path to the project directory.
 * - `importPath`: The import path for the library.
 * - `appsDir`: The path to the apps directory.
 * - `libsDir`: The path to the libs directory.
 * - `parsedTags`: The parsed tags for the project.
 * - `npmScope`: The npm scope for the library.
 * - `npmScopeTitle`: The title cased npm scope for the library.
 * - `offsetFromRoot`: The relative path from the workspace root to the library.
 * - `loggingLibName`: The name of the logging library.
 * - `loggingLibImportPath`: The import path for the logging library.
 * - `typesLibName`: The name of the types library.
 * - `typesLibImportPath`: The import path for the types library.
 */
export function normalizeOptions(
  tree: Tree,
  options: DbGeneratorSchema,
  extra: Pick<LoggingLibNormalizedSchema, 'projectName' | 'importPath' | 'typesLibName' | 'typesLibImportPath'>,
): NormalizedSchema {
  const layout = getWorkspaceLayout(tree);
  const name = names(options.name ?? defaultLibName).fileName;
  const project = names(defaultLibDirectory).fileName;
  const appsDir = layout.appsDir === '.' ? 'apps' : layout.appsDir;
  const libsDir = layout.libsDir === '.' ? 'libs' : layout.libsDir;
  const projectDirectory = `${project}/${name}`;
  const projectRoot = `${libsDir}/${projectDirectory}`;
  const importPath = getImportPath(tree, name);
  const npmScope = getNpmScope(tree) ?? name;
  const npmScopeTitle = names(npmScope).className;
  const offsetFromRootPath = offsetFromRoot(projectRoot);

  return {
    ...options,
    appsDir,
    importPath,
    libsDir,
    name: options.name,
    parsedTags: defaultLibTags,
    projectDirectory,
    projectName: name,
    projectRoot,
    npmScope,
    npmScopeTitle,
    offsetFromRoot: offsetFromRootPath,
    loggingLibName: extra.projectName,
    loggingLibImportPath: extra.importPath,
    typesLibName: extra.typesLibName,
    typesLibImportPath: extra.typesLibImportPath,
  };
}
