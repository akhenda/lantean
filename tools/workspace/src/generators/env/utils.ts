import { getWorkspaceLayout, names, Tree } from '@nx/devkit';

import { getImportPath } from '../../utils';

import { defaultLibDirectory, defaultLibName, defaultLibTags } from './constants';
import { NormalizedSchema, EnvGeneratorSchema } from './schema';

/**
 * Normalize options for the ESLint generator.
 *
 * @param tree The virtual file system tree.
 * @param options The options passed to the generator.
 * @returns The normalized options.
 *
 * The normalized options include the following:
 * - `name`: The name of the ESLint config library.
 * - `projectName`: The name of the project.
 * - `projectRoot`: The path to the project root.
 * - `projectDirectory`: The path to the project directory.
 * - `parsedTags`: The parsed tags for the project.
 * - `appsDir`: The path to the apps directory.
 * - `libsDir`: The path to the libs directory.
 * - `importPath`: The import path for the ESLint config library.
 */
export function normalizeOptions(
  tree: Tree,
  options: EnvGeneratorSchema = { name: defaultLibName },
): NormalizedSchema {
  const layout = getWorkspaceLayout(tree);
  const name = names(options.name).fileName;
  const project = names(defaultLibDirectory).fileName;
  const appsDir = layout.appsDir === '.' ? 'apps' : layout.appsDir;
  const libsDir = layout.libsDir === '.' ? 'libs' : layout.libsDir;
  const projectDirectory = `${project}/${name}`;
  const projectRoot = `${libsDir}/${projectDirectory}`;

  return {
    ...options,
    appsDir,
    importPath: getImportPath(tree, name),
    libsDir,
    name: options.name,
    parsedTags: defaultLibTags,
    projectDirectory,
    projectName: name,
    projectRoot,
  };
}
