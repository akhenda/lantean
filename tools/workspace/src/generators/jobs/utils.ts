import { getWorkspaceLayout, names, offsetFromRoot, Tree } from '@nx/devkit';

import { defaultLibDirectory, defaultLibName, defaultLibTags } from './constants';
import { NormalizedSchema, JobsGeneratorSchema } from './schema';

import { getImportPath, getNpmScope } from '../../utils';

/**
 * Normalize options for the Jobs generator.
 *
 * @param tree The virtual file system tree.
 * @param options The options passed to the generator.
 * @param extra The extra options passed to the generator.
 * @returns The normalized options.
 *
 * The normalized options include the following:
 * - `name`: The name of the library.
 * - `projectName`: The name of the project.
 * - `projectRoot`: The path to the project root.
 * - `projectDirectory`: The path to the project directory.
 * - `importPath`: The import path for the library.
 * - `appsDir`: The path to the apps directory.
 * - `libsDir`: The path to the libs directory.
 * - `parsedTags`: The parsed tags for the project.
 * - `npmScope`: The npm scope for the library.
 * - `npmScopeTitle`: The title cased npm scope for the library.
 * - `offsetFromRoot`: The relative path from the workspace root to the library.
 */
export function normalizeOptions(tree: Tree, options: JobsGeneratorSchema): NormalizedSchema {
  const layout = getWorkspaceLayout(tree);
  const nameOptions = names(options.name ?? defaultLibName);
  const name = nameOptions.fileName;
  const project = names(defaultLibDirectory).fileName;
  const appsDir = layout.appsDir === '.' ? 'apps' : layout.appsDir;
  const libsDir = layout.libsDir === '.' ? 'libs' : layout.libsDir;
  const projectDirectory = `${project}/${name}`;
  const projectRoot = `${libsDir}/${projectDirectory}`;
  const importPath = getImportPath(tree, name);
  const npmScope = getNpmScope(tree) ?? name;
  const npmScopeTitle = names(npmScope).className;
  const offsetFromRootPath = offsetFromRoot(projectRoot);

  return {
    ...options,
    appsDir,
    importPath,
    libsDir,
    name: nameOptions.name,
    parsedTags: defaultLibTags,
    projectDirectory,
    projectName: name,
    projectRoot,
    npmScope,
    npmScopeTitle,
    offsetFromRoot: offsetFromRootPath,
  };
}
