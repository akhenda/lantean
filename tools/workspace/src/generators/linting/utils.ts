import { getWorkspaceLayout, names, Tree, updateJson } from '@nx/devkit';
import { JSONSchemaForTheTypeScriptCompilerSConfigurationFile as TSConfig } from '@schemastore/tsconfig';

import { getImportPath, updateTSConfigCompilerOptions } from '../../utils';

import { eslintLibDirectory, eslintLibName, eslintLibTags } from './constants';
import { LintingGeneratorSchema } from './schema';
import { NormalizedSchema } from './types';

/**
 * Normalize options for the ESLint generator.
 *
 * @param tree The virtual file system tree.
 * @param options The options passed to the generator.
 * @returns The normalized options.
 *
 * The normalized options include the following:
 * - `name`: The name of the ESLint config library.
 * - `projectName`: The name of the project.
 * - `projectRoot`: The path to the project root.
 * - `projectDirectory`: The path to the project directory.
 * - `parsedTags`: The parsed tags for the project.
 * - `appsDir`: The path to the apps directory.
 * - `libsDir`: The path to the libs directory.
 * - `importPath`: The import path for the ESLint config library.
 */
export function normalizeOptions(
  tree: Tree,
  options: LintingGeneratorSchema = {},
): NormalizedSchema {
  const layout = getWorkspaceLayout(tree);
  const name = names(eslintLibName).fileName;
  const project = names(eslintLibDirectory).fileName;
  const appsDir = layout.appsDir === '.' ? 'apps' : layout.appsDir;
  const libsDir = layout.libsDir === '.' ? 'libs' : layout.libsDir;
  const projectDirectory = `${project}/${name}`;
  const projectRoot = `${libsDir}/${projectDirectory}`;
  const importPath = getImportPath(tree, name);

  return {
    ...options,
    appsDir,
    importPath,
    libsDir,
    name: eslintLibName,
    parsedTags: eslintLibTags,
    projectDirectory,
    projectName: name,
    projectRoot,
  };
}

/**
 * Updates the base TSConfig.
 *
 * - Enables strictNullChecks in the base TSConfig.
 *
 * @param tree The virtual file system tree.
 * @param options The normalized schema options.
 */
export function updateBaseTSConfig(tree: Tree) {
  updateJson<TSConfig>(tree, 'tsconfig.base.json', (json) => {
    return updateTSConfigCompilerOptions(json, { strictNullChecks: true });
  });
}
