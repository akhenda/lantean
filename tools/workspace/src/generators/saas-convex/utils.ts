import { getWorkspaceLayout, Tree } from '@nx/devkit';

import { NormalizedSchema, SaasConvexGeneratorSchema } from './schema';

import { normalizeOptions as normalizeInitOptions } from '../init/utils';
import { normalizeOptions as normalizeCommitlintOptions } from '../commitlint/utils';
import { normalizeOptions as normalizeSheriffOptions } from '../sheriff/utils';
import { normalizeOptions as normalizeEnvOptions } from '../env/utils';
import { normalizeOptions as normalizeTypesOptions } from '../types/utils';
import { normalizeOptions as normalizeLoggingOptions } from '../logging/utils';
import { normalizeOptions as normalizeMonitoringOptions } from '../monitoring/utils';
import { normalizeOptions as normalizeAnalyticsOptions } from '../analytics/utils';
import { normalizeOptions as normalizeDateOptions } from '../date/utils';
import { normalizeOptions as normalizeEmailOptions } from '../email/utils';
import { normalizeOptions as normalizeJobsOptions } from '../jobs/utils';
import { normalizeOptions as normalizeKVOptions } from '../kv/utils';
import { normalizeOptions as normalizeConvexOptions } from '../convex/utils';
import { normalizeOptions as normalizeUniversalOptions } from '../universal/utils';
import { normalizeOptions as normalizeUniversalExpoOptions } from '../universal-expo/utils';
import { normalizeOptions as normalizeUniversalNextOptions } from '../universal-next/utils';
import { defaultUniversalLibLibName, defaultUniversalLibUIName } from './constants';

/**
 * Normalize options for the SaasConvex generator.
 *
 * @param tree The virtual file system tree.
 * @param options The options passed to the generator.
 * @returns The normalized options.
 *
 * The normalized options include the following:
 * - `appsDir`: The path to the apps directory.
 * - `libsDir`: The path to the libs directory.
 * - `lintStagedConfigFileName`: The file name for the lint-staged config file.
 * - `commitLintConfigFileName`: The file name for the commitlint config file.
 * - `envLibName`: The name of the env library.
 * - `typesLibName`: The name of the types library.
 * - `loggingLibName`: The name of the logging library.
 * - `monitoringLibName`: The name of the monitoring library.
 * - `analyticsLibName`: The name of the analytics library.
 * - `dateLibName`: The name of the date library.
 * - `emailLibName`: The name of the email library.
 * - `jobsLibName`: The name of the jobs library.
 * - `kvLibName`: The name of the KV library.
 * - `convexLibName`: The name of the Convex library.
 * - `universalLibUIName`: The name of the universal library for the UI.
 * - `universalLibLibName`: The name of the universal library for utils.
 * - `expoAppName`: The name of the Expo app.
 * - `expoAppDisplayName`: The display name of the Expo app.
 * - `nextJSAppName`: The name of the Next.js app.
 */
export function normalizeOptions(tree: Tree, { skipFormat, ...options }: SaasConvexGeneratorSchema): NormalizedSchema {
  const layout = getWorkspaceLayout(tree);
  const appsDir = layout.appsDir === '.' ? 'apps' : layout.appsDir;
  const libsDir = layout.libsDir === '.' ? 'libs' : layout.libsDir;

  const { fileName: lintStagedConfigFileName } = normalizeInitOptions({
    configFileName: options.lintStagedConfigFileName,
  });
  const { fileName: commitLintConfigFileName } = normalizeCommitlintOptions(tree, {
    configFileName: options.commitLintConfigFileName,
  });
  const { importPath: sheriffImportPath } = normalizeSheriffOptions(tree);
  const { projectName: envLibName } = normalizeEnvOptions(tree, {
    name: options.envLibName,
  });
  const { projectName: typesLibName, importPath: typesImportPath } = normalizeTypesOptions(tree, {
    name: options.typesLibName,
  });
  const { projectName: loggingLibName, importPath: loggingImportPath } = normalizeLoggingOptions(
    tree,
    { name: options.loggingLibName, typesLibName, skipFormat },
    { projectName: typesLibName, importPath: typesImportPath },
  );
  const { projectName: monitoringLibName } = normalizeMonitoringOptions(
    tree,
    { name: options.monitoringLibName, typesLibName, skipFormat },
    { projectName: typesLibName, importPath: typesImportPath },
  );
  const { projectName: analyticsLibName } = normalizeAnalyticsOptions(
    tree,
    { name: options.analyticsLibName, typesLibName, skipFormat, loggingLibName },
    { projectName: loggingLibName, importPath: loggingImportPath, typesLibName, typesLibImportPath: typesImportPath },
  );
  const { projectName: dateLibName } = normalizeDateOptions(tree, {
    name: options.dateLibName,
  });
  const { projectName: emailLibName } = normalizeEmailOptions(tree, {
    name: options.emailLibName,
  });
  const { projectName: jobsLibName } = normalizeJobsOptions(tree, {
    name: options.jobsLibName,
  });
  const { projectName: kvLibName } = normalizeKVOptions(tree, {
    name: options.kvLibName,
  });
  const { projectName: convexLibName } = normalizeConvexOptions(tree, {
    name: options.convexLibName,
  });
  const { uiName: universalLibUIName, libName: universalLibLibName } = normalizeUniversalOptions(tree, {
    uiName: options.universalLibUIName ?? defaultUniversalLibUIName,
    libName: options.universalLibLibName ?? defaultUniversalLibLibName,
    skipFormat,
  });
  const { projectName: expoAppName, displayName: expoAppDisplayName } = normalizeUniversalExpoOptions(tree, {
    name: options.expoAppName,
    displayName: options.expoAppDisplayName,
    uiName: universalLibUIName,
    libName: universalLibLibName,
  });
  const { projectName: nextJSAppName } = normalizeUniversalNextOptions(tree, {
    name: options.nextJSAppName,
    uiName: universalLibUIName,
    libName: universalLibLibName,
  });

  return {
    ...options,
    appsDir,
    libsDir,
    lintStagedConfigFileName,
    commitLintConfigFileName,
    sheriffImportPath,
    envLibName,
    typesLibName,
    loggingLibName,
    monitoringLibName,
    analyticsLibName,
    dateLibName,
    emailLibName,
    jobsLibName,
    kvLibName,
    convexLibName,
    universalLibUIName,
    universalLibLibName,
    expoAppName,
    expoAppDisplayName,
    nextJSAppName,
  };
}
