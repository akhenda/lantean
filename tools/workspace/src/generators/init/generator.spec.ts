import { Tree, installPackagesTask, readJson, writeJson } from '@nx/devkit';
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';

import {
  addDevDependencyToPackageJson,
  addHuskyHook,
  addHuskyToPackageJson,
  installHuskyTask,
} from '../../devkit';
import * as utils from '../../utils';

import lintStagedGenerator from './generator';

jest.mock('@nx/devkit', () => ({
  ...jest.requireActual('@nx/devkit'),
  installPackagesTask: jest.fn(),
}));

jest.mock('../../devkit', () => ({
  ...jest.requireActual('../../devkit'),
  addDevDependencyToPackageJson: jest.fn(),
  addHuskyToPackageJson: jest.fn(),
  addHuskyHook: jest.fn(),
  installHuskyTask: jest.fn(),
}));

describe('@nx-squeezer/workspace lint-staged generator', () => {
  let tree: Tree;

  beforeEach(() => {
    tree = createTreeWithEmptyWorkspace();
    jest.spyOn(console, 'log').mockImplementation(() => null);
    jest.clearAllMocks();
    jest.restoreAllMocks();
  });

  it('should run successfully', async () => {
    await lintStagedGenerator(tree);
  });

  it('should add dependencies', async () => {
    await lintStagedGenerator(tree);

    expect(addHuskyToPackageJson).toHaveBeenCalled();
    expect(addDevDependencyToPackageJson).toHaveBeenCalledWith(
      tree,
      'lint-staged'
    );
  });

  it('should add husky hook', async () => {
    jest.spyOn(utils, 'isHuskyInstalled').mockReturnValueOnce(true);

    await lintStagedGenerator(tree);

    expect(addHuskyHook).toHaveBeenCalledWith(
      tree,
      'pre-commit',
      'npx lint-staged --concurrent false --relative'
    );
  });

  it('should create lint-staged config', async () => {
    const configFileName = '.lintstagedrc';

    await lintStagedGenerator(tree, { configFileName });

    expect(readJson(tree, configFileName)).toMatchSnapshot();
  });

  it('should merge lint-staged config', async () => {
    const configFileName = '.lintstagedrc';
    const mdFormat = { '*.md': 'markdown --format' };

    writeJson(tree, configFileName, mdFormat);

    await lintStagedGenerator(tree, { configFileName });

    expect(readJson(tree, configFileName)).toMatchSnapshot();
  });

  it('should run tasks', async () => {
    jest.spyOn(utils, 'isHuskyInstalled').mockReturnValueOnce(false);

    const tasks = await lintStagedGenerator(tree);

    expect(tasks).toBeTruthy();

    tasks?.();

    expect(utils.isHuskyInstalled).toHaveBeenCalled();
    expect(installPackagesTask).toHaveBeenCalled();
    expect(installHuskyTask).toHaveBeenCalled();
    expect(addHuskyHook).not.toHaveBeenCalled();
  });
});
