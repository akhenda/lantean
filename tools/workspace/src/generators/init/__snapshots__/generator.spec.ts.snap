// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@nx-squeezer/workspace lint-staged generator should create lint-staged config 1`] = `
{
  "{apps,libs,packages,tools}/**/*.{js,ts,jsx,tsx,json,md,html,css,scss}": [
    "nx affected --target=lint --fix",
    "nx format:write",
  ],
  "{apps,libs,tools}/**/*.{ts,tsx}": [
    "nx affected --target=typecheck",
  ],
}
`;

exports[`@nx-squeezer/workspace lint-staged generator should merge lint-staged config 1`] = `
{
  "*.md": "markdown --format",
}
`;
